### PR TITLE
[ROCm] XLA unit-test updates for the ROCm platform

### PR DESCRIPTION
**Pull request by deven-amd**
This PR has commits which update XLA unit-tests for the ROCm platform.
* Few of the testscases need ROCm specific versions for the expected results for FileCheck (since the IR generated for AMD GPUs is different than the same for CUDA based GPUs)
* adding no_rocm tag to tests that currently fail in the upstream ROCm TF version

--------------------

/cc @whchung @chsigg

### CODE REVIEW

File: tensorflow/compiler/xla/tests/filecheck.cc [hunk 1]
```diff
@@ -40,7 +40,7 @@ StatusOr<bool> RunFileCheck(const std::string& input,
 
   // Invoke FileCheck to check whether input matches `pattern`.
   const char* file_check_path_suffix =
-      "org_tensorflow/external/llvm/FileCheck";
+      "org_tensorflow/external/llvm-project/llvm/FileCheck";
```
Comment (cheshire) [inline tensorflow/compiler/xla/tests/filecheck.cc]: Why this change?

File: tensorflow/compiler/xla/service/gpu/tests/gpu_ldg_test.cc [hunk 2]
```diff
@@ -38,6 +38,11 @@ class GpuLdgTest : public GpuCodegenTest {};
 
 // Parameters are never overwritten, so parameter reads should get ld.global.nc
 // reads.
+//
+// On the ROCM platform the "ptx" string is not populated for the compiled
```
Comment (cheshire) [inline tensorflow/compiler/xla/service/gpu/tests/gpu_ldg_test.cc]: Should we just disable these tests for ROCM?

File: tensorflow/compiler/xla/service/gpu/tests/gpu_kernel_tiling_test.cc [hunk 3]
```diff
@@ -602,11 +769,19 @@ TEST_F(GpuKernelTilingTest, RowReductionWithSmallDimensionNotTiled) {
       ParseAndReturnVerifiedModule(kHloString, ConfigWithoutLayoutAssignment())
           .ValueOrDie();
   CompileAndVerifyIr(std::move(hlo_module),
+#if TENSORFLOW_USE_ROCM
```
Comment (cheshire) [inline tensorflow/compiler/xla/service/gpu/tests/gpu_kernel_tiling_test.cc]: Could we at runtime instead of an ifdef here?

File: tensorflow/compiler/tests/build_defs.bzl [hunk 4]
```diff
@@ -85,10 +86,16 @@ def tf_xla_py_test(
...
-                "--test_device=" + gpu_xla_device,
-                "--types=DT_HALF,DT_FLOAT,DT_DOUBLE,DT_UINT8,DT_QUINT8,DT_INT8,DT_QINT8,DT_INT32,DT_QINT32,DT_INT64,DT_BOOL,DT_COMPLEX64,DT_COMPLEX128,DT_BFLOAT16",
-            ]
+            if rocm_is_configured():
+                backend_args += [
+                    "--test_device=" + gpu_xla_device,
+                    "--types=DT_HALF,DT_FLOAT,DT_DOUBLE,DT_UINT8,DT_QUINT8,DT_INT8,DT_QINT8,DT_INT32,DT_QINT32,DT_INT64,DT_BOOL,DT_COMPLEX64,DT_BFLOAT16",
+                ]
+            else:
+                backend_args += [
```
Comment (cheshire) [inline tensorflow/compiler/tests/build_defs.bzl]: Is it possible to avoid the duplication in these cases? Currently it's hard to see what is the difference between the cases.

**Review by cheshire — CHANGES_REQUESTED** [review]
Mostly looks great, thanks!

File: tensorflow/compiler/xla/tests/filecheck.cc (hunk 1 above)
Comment (deven-amd) [reply tensorflow/compiler/xla/tests/filecheck.cc]: It is required as a consequence of this commit : https://github.com/tensorflow/tensorflow/commit/d6ba353dd974f3d883734e96653cdee7fe6abfbc

The path of the`FileCheck` executable changed after that commit, and that change is reflected here.

File: tensorflow/compiler/xla/service/gpu/tests/gpu_ldg_test.cc (hunk 2 above)
Comment (deven-amd) [reply tensorflow/compiler/xla/service/gpu/tests/gpu_ldg_test.cc]: We still wanted to keep the "Compile" part of the "CompileAndVerifyPtx" and hence the comment, instead of disabling it altogether

something is better than nothing :)

File: tensorflow/compiler/xla/service/gpu/tests/gpu_kernel_tiling_test.cc (hunk 3 above)
Comment (deven-amd) [reply tensorflow/compiler/xla/service/gpu/tests/gpu_kernel_tiling_test.cc]: guess we could try and make this a runtime check instead of an ifdef, but what would be gain by going that route?

Making this a compile time ifdef, does make the code easier to follow.

File: tensorflow/compiler/tests/build_defs.bzl (hunk 4 above)
Comment (deven-amd) [reply tensorflow/compiler/tests/build_defs.bzl]: good point.

I will make this change and push out the commit

Comment (deven-amd) [reply tensorflow/compiler/tests/build_defs.bzl]: removed the special casing for rocm altogether. since the PR to switch to ROCm 3.0 was recently merged, this is no longer needed :)

**Review by deven-amd — COMMENTED** [review]
@cheshire , I have rebased the PR to remove the merge conflict and addressed the code-review comments. 

Please re-review.

thanks

File: tensorflow/compiler/xla/tests/filecheck.cc (hunk 1 above)
Comment (cheshire) [reply tensorflow/compiler/xla/tests/filecheck.cc]: Right, but why the tests are passing now without this change?

Comment (deven-amd) [reply tensorflow/compiler/xla/tests/filecheck.cc]: They are not passing on our end without this change. I do not have visibility into your end

File: tensorflow/compiler/xla/service/gpu/tests/gpu_kernel_tiling_test.cc (hunk 3 above)
Comment (cheshire) [reply tensorflow/compiler/xla/service/gpu/tests/gpu_kernel_tiling_test.cc]: I'd still prefer a runtime check, if possible, as in future we might want to switch to shipping both ROCM and CUDA in the same binary.

File: tensorflow/compiler/xla/tests/filecheck.cc (hunk 1 above)
Comment (cheshire) [reply tensorflow/compiler/xla/tests/filecheck.cc]: > I do not have visibility into your end

You are right that tests are run in a different environment inside Google. However, presubmits which are being run by Github bot externally should be run inside the environment identical to yours, and you sort of do have visibility into it, as the bot should log all the actions it performs and should start from a reproducible state.

In any case, if the tests don't pass for you without this change, could you file a separate PR with this change only, and we could discuss/check it there?

**Review by cheshire — CHANGES_REQUESTED** [review]
If you need a change to FileCheck path, could you please file a separate PR for it? Our open-source tests run in an environment which should be (hopefully almost) identical to yours, so it should not be the case that one set of tests failing and another is not.

Additionally, if possible, I would still prefer runtime switch in tests.
If anything, it would make sure that we always compile both branches, and would help to avoid compilation breakages due to e.g. refactoring.

File: tensorflow/compiler/xla/tests/filecheck.cc (hunk 1 above)
Comment (deven-amd) [reply tensorflow/compiler/xla/tests/filecheck.cc]: filed PR https://github.com/tensorflow/tensorflow/pull/35751 as per your reqeust

File: tensorflow/compiler/xla/service/gpu/tests/gpu_kernel_tiling_test.cc (hunk 3 above)
Comment (deven-amd) [reply tensorflow/compiler/xla/service/gpu/tests/gpu_kernel_tiling_test.cc]: Thought about this some more, and I do not know whether this is feasible.

Currently we can use the stream executor to do a Runtime check to determine whether we are using ROCm or CUDA. (see example here : https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/service/gpu/gpu_conv_algorithm_picker.cc#L322-L326 )

How do I get access to a stream executor object within this test object?

**Comment by cheshire** [comment]
There are global getters. You could either do
`PlatformUtil::GetDefaultPlatform()` and try to get a stream executor from
there (e.g.
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/tests/while_test.cc#L1267),
or check whether `se::MultiPlatformManager::PlatformWithName` returns a
platform for a given name (cf.
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/service/gpu/buffer_comparator_test.cc#L35
)

File: tensorflow/compiler/xla/service/gpu/tests/gpu_kernel_tiling_test.cc (hunk 3 above)
Comment (cheshire) [reply tensorflow/compiler/xla/service/gpu/tests/gpu_kernel_tiling_test.cc]: I think I have replied using an email, but just in case it didn't make it:

There are global getters. You could either do `PlatformUtil::GetDefaultPlatform()` and try to get a stream executor from there (e.g. https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/tests/while_test.cc#L1267), or check whether `se::MultiPlatformManager::PlatformWithName` returns a platform for a given name (cf. https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/service/gpu/buffer_comparator_test.cc#L35)

Having a runtime switch has another advantage that both branches would always be compiled, and editing one would not accidentally break the compilation of another.

Comment (deven-amd) [reply tensorflow/compiler/xla/service/gpu/tests/gpu_kernel_tiling_test.cc]: @cheshire 

I have have pushed out a new commit, which makes the expected_ir switch at runtime. Please re-review.  thanks

deven

File: tensorflow/compiler/xla/service/gpu/tests/gpu_codegen_test.cc [hunk 5]
```diff
@@ -51,9 +51,15 @@ void GpuCodegenTest::CompileAndVerifyPtx(
   std::unique_ptr<Executable> executable =
       std::move(CompileToExecutable(std::move(hlo_module)).ValueOrDie());
   string ptx_str(static_cast<GpuExecutable*>(executable.get())->text());
+
+  // On the ROCM platform the "ptx" string is not populated for the compiled
+  // executable, and hence the "ptx_str" will be empty. So disabling the
+  // pattern check on the ROCm platform
+#if !defined(TENSORFLOW_USE_ROCM)
```
Comment (cheshire) [inline tensorflow/compiler/xla/service/gpu/tests/gpu_codegen_test.cc]: Sorry, same comment as before, would it be possible to make it a runtime check?

Comment (cheshire) [reply tensorflow/compiler/xla/service/gpu/tests/gpu_codegen_test.cc]: You even have a `is_built_with_rocm_` variable now

File: tensorflow/compiler/xla/service/gpu/tests/gpu_ftz_test.cc [hunk 6]
```diff
@@ -76,13 +76,23 @@ class GpuFtzDisabledTest : public GpuFtzTest {
 };
 
 // Check that we emit mul.ftz.f32 when in ftz mode, and plain mul.f32 otherwise.
+//
+// On the ROCM platform the "ptx" string is not populated for the compiled
+// executable, and hence the call to CompileAdnVerifyPtx does not do the
+// "VerifyPtx" part, it merely compiles the executable
```
Comment (cheshire) [inline tensorflow/compiler/xla/service/gpu/tests/gpu_ftz_test.cc]: Typo: CompileAndVerifyPtx.

Are we testing anything in ROCM mode here at all? Would you like to just test the compilation?

Currently the API creates a false impression of testing on ROCM which ideally would be good to avoid.

You could also have a test-local function e.g. `CompileAndOptionallyVerifyPtx`, which would verify PTX only if running on CUDA?

File: tensorflow/compiler/xla/service/gpu/tests/gpu_index_test.cc [hunk 7]
```diff
@@ -105,13 +105,24 @@ TEST_F(GpuIndexTest, CompatibleUseLinearIndexWithReshapeAndBroadcast) {
...
-  CompileAndVerifyIr(std::move(module),
-                     R"(
+
+  // In the IR generated for AMDGPUs, we do not seem to have the
+  // the addrspace(1) attribute for the lines being checked by the following
+  // patterns still need to investigate why that is the case, and whether or not
+  // it is ok
+  auto expected_ir = is_built_with_rocm_ ? R"(
+; CHECK: %[[urem1:.*]] = urem i{{[0-9]*}} %[[linear_index:.*]], 14
+; CHECK: %[[bitcast:.*]] = bitcast i8* %[[alloc:.*]] to float*
```
Comment (cheshire) [inline tensorflow/compiler/xla/service/gpu/tests/gpu_index_test.cc]: It's probably not worth it to duplicate the pattern to catch addrspace cast. Could you use a regexp to optionally match and addrspace cast instead?

File: tensorflow/compiler/xla/service/gpu/tests/gpu_input_fusible_slice_test.cc [hunk 8]
```diff
@@ -63,12 +63,17 @@ TEST_F(GpuSliceInputFusionTest, InputFusionWithOnlyOneSlice) {
   auto hlo_module =
       ParseAndReturnVerifiedModule(kHloString, ConfigWithoutLayoutAssignment())
           .ValueOrDie();
-  CompileAndVerifyIr(std::move(hlo_module),
-                     R"(
+  auto expected_ir = is_built_with_rocm_ ? R"(
+; CHECK-LABEL: define amdgpu_kernel void @fusion
```
Comment (cheshire) [inline tensorflow/compiler/xla/service/gpu/tests/gpu_input_fusible_slice_test.cc]: Could we just optionally match amdgpu_kernel using a regexp instead of duplication?

Comment (cheshire) [reply tensorflow/compiler/xla/service/gpu/tests/gpu_input_fusible_slice_test.cc]: Same applies below.

**Review by cheshire — CHANGES_REQUESTED** [review]
Mostly looks great, sorry for all the nitpicks!

File: tensorflow/compiler/xla/service/gpu/tests/gpu_codegen_test.cc (hunk 5 above)
Comment (deven-amd) [reply tensorflow/compiler/xla/service/gpu/tests/gpu_codegen_test.cc]: will fix this in the next commit I push out

File: tensorflow/compiler/xla/service/gpu/tests/gpu_ftz_test.cc (hunk 6 above)
Comment (deven-amd) [reply tensorflow/compiler/xla/service/gpu/tests/gpu_ftz_test.cc]: in the next commit, I will rename the existing `CompileAndVerifyPtx` routine to `CompileAndOptionalyVerifyPtx`, add comments in the `gpu_codegen_test.[h,cc]` indicating why, and also remove all instances of the comment above

File: tensorflow/compiler/xla/service/gpu/tests/gpu_input_fusible_slice_test.cc (hunk 8 above)
Comment (deven-amd) [reply tensorflow/compiler/xla/service/gpu/tests/gpu_input_fusible_slice_test.cc]: this may not be viable.

We want to check for presence of `amdgpu_kernel` in ROCm mode, so having a regexp that optionally matches `amdgpu_kernel` will not work, since it will pass even if `amdgpu_kernel` is not present!

can we leave this as is?

File: tensorflow/compiler/xla/service/gpu/tests/gpu_index_test.cc (hunk 7 above)
Comment (deven-amd) [reply tensorflow/compiler/xla/service/gpu/tests/gpu_index_test.cc]: will push this change out in the next commit as well.

**Review by deven-amd — COMMENTED** [review]
@cheshire , I have pushed out a commit with changes you requested...please re-review. thanks

File: tensorflow/compiler/xla/service/gpu/tests/gpu_input_fusible_slice_test.cc (hunk 8 above)
Comment (cheshire) [reply tensorflow/compiler/xla/service/gpu/tests/gpu_input_fusible_slice_test.cc]: OK sure, in general we have to weigh in the risk of one of the branches getting stale (recall that currently we do not run ROCM tests internally, so it's possible that someone would update the nvidia branch, but not the ROCM branch) vs. the exact pattern you want to match. So here I would say that matching `amdgpu_kernel` is not important enough to have two branches, but since the pattern is small I don't care that much either.

File: tensorflow/compiler/xla/service/gpu/tests/gpu_kernel_tiling_test.cc [hunk 9]
```diff
@@ -424,16 +495,43 @@ TEST_F(GpuKernelTilingTest, ColumnReductionMOFUnrolled) {
...
           .ValueOrDie();
-  CompileAndVerifyIr(std::move(hlo_module),
-                     R"(
+  auto expected_ir = is_built_with_rocm_ ? R"(
+; CHECK-LABEL: define amdgpu_kernel void @fusion
+;
+; CHECK-LABEL: atomic_op_loop_body{{.*}}:
+; CHECK: %[[fadd:.*]] = fadd float %{{.*}}, %{{.*}}
+; CHECK: %[[bitcast:.*]] = bitcast float %[[fadd]] to i32
+; CHECK: %{{.*}} = cmpxchg i32* %{{.*}}, i32 %{{.*}}, i32 %[[bitcast]]
```
Comment (cheshire) [inline tensorflow/compiler/xla/service/gpu/tests/gpu_kernel_tiling_test.cc]: Optional: I think the important bit is that we have 4 `cmpxchg` instructions, I would even match only those.

**Review by cheshire — APPROVED** [review]
LGTM with minor nits.

File: tensorflow/compiler/xla/service/gpu/tests/gpu_input_fusible_slice_test.cc (hunk 8 above)
Comment (deven-amd) [reply tensorflow/compiler/xla/service/gpu/tests/gpu_input_fusible_slice_test.cc]: @whchung , what say you?

If you are okay with it, I can push out new PR which makes the check for `amdgpu_kernel` optional?

**Review by cheshire — APPROVED** [review]
